sycl/include/sycl/ext/oneapi/bfloat16.hpp

```diff
@@ -30,16 +34,31 @@ class bfloat16 {
   bfloat16(const bfloat16 &) = default;
   ~bfloat16() = default;
 
+private:
   // Explicit conversion functions
   static storage_t from_float(const float &a) {
 #if defined(__SYCL_DEVICE_ONLY__)
 #if defined(__NVPTX__)
+#if (__CUDA_ARCH__ >= 800)
     return __nvvm_f2bf16_rn(a);
 #else
-    return __spirv_ConvertFToBF16INTEL(a);
+    // TODO std::isnan not defined in device code
+    // if (std::isnan(a))
+    // return 0xffc1;
```

Comment:
This is still commented out.  Can you use the same code you have in devicelib?

Reply:
This isnan usage is in the CUDA code. I will leave this to be done later.

Reply:
@JackAKirk do you have any suggestion here?  Is it OK to leave this as TODO, or does it need to be fixed for the next compiler release?

Reply:
a != a is unlikely to survive compiler optimization, but I will make that change until we have something better.

Reply:
Clang is very cautious regarding optimizing floating point operations, and for example here https://godbolt.org/z/nxhbezr11 `fcmp` survives O2 (clang even adds `uno` argument aka check for NaN). But yeah, I remember facing a sporadic accuracy-related bug due caused by the fact, that MSVC is much less careful in such scenarios, optimizing floating point arithmetic where it shouldn't.

Reply:
It would be good to add a test case checking the nan behavior of `from_float`. Also I think the bfloat16_type.cpp test should actually be set to run (rather than just check the test can be compiled) on all currently supported backends before taking bfloat16 out of the experimental namespace. If the CI machines support the native implementations then we should be able to compile and run tests for both native and software implementations. For the CUDA backend the CI does not have a machine supporting the native implementation yet, but we can still check that it compiles in that case: For example see: https://github.com/rdeodhar/llvm-test-suite/pull/1/files

Reply:
I've added a test for bfloat16 conversions in intel/llvm-test-suite, bfloat16_conversions.cpp. See PR 1129 in intel/llvm-test-suite.
Regarding the existing test bfloat16_types.cpp, it can now be run on CPU and GPU and will use native or emulated bfloat16 support depending on GPU found at runtime.